### PR TITLE
[SourceKit] Preparation for solver-based cursor info

### PR DESCRIPTION
**Pull request by ahoppen**
This PR contains multiple, more or less independent commits that are required to implement cursor info using the solver-based / completion-like design that doesn’t type check the entire file but only those parts that are necessary to server the request and which will allow cursor info to return ambiguous results.

If any of these changes is controversial, I can split it into a separate PR.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

File: include/swift/IDE/Utils.h [hunk 1]
```diff
@@ -128,71 +128,201 @@ enum class CursorInfoKind {
...
+    ExtensionDecl *ExtTyRef = nullptr;
+    bool IsRef = true;
+    Type Ty;
+    Type ContainerType;
+    bool IsKeywordArgument = false;
+    /// It this is a ref, whether it is "dynamic". See \c ide::isDynamicRef.
+    bool IsDynamic = false;
+    /// If this is a dynamic ref, the types of the base (multiple in the case of
+    /// protocol composition).
+    SmallVector<NominalTypeDecl *, 1> ReceiverTypes;
```
Comment (bnbarham) [inline include/swift/IDE/Utils.h:158]: While you're here you may as well remove the default from both of these and let `SmallVector` use whatever fits instead.

File: lib/IDE/IDETypeChecking.cpp [hunk 2]
```diff
@@ -915,13 +915,21 @@ Type swift::getResultTypeOfKeypathDynamicMember(SubscriptDecl *SD) {
...
     }
-    auto *DRE = dyn_cast_or_null<DeclRefExpr>(Capture.PBD->getInit(0));
+    Expr *Init = Capture.PBD->getInit(0);
+    if (!Init) {
+      continue;
+    }
+    if (isa<UnresolvedDeclRefExpr>(Init) && DC) {
+      // Type check Init to resolve the decl ref.
+      typeCheckExpression(DC, Init);
+    }
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply lib/IDE/IDETypeChecking.cpp:931]: Doing some more work on this, I changed this to
```cpp
Init = resolveDeclRefExpr(UDRE, DC, /*replaceInvalidRefsWithErrors=*/false);
```
which should be a lot faster and seems reasonable to me.

File: include/swift/IDE/Utils.h (hunk 1 above)
Comment (ahoppen) [reply include/swift/IDE/Utils.h:158]: I forgot that `SmallVector` now has a default for its size.

File: include/swift/IDE/Utils.h [hunk 3]
```diff
@@ -128,71 +128,201 @@ enum class CursorInfoKind {
   StmtStart,
 };
 
+/// Base class of more specialized \c ResolvedCursorInfos that also represents
+/// and \c Invalid cursor info.
+/// Subclasses of \c ResolvedCursorInfo cannot add new stored properies because
+/// \c ResolvedCursorInfo is being passed around as its base class and thus any
+/// properties in subclasses would get lost.
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply include/swift/IDE/Utils.h:135]: I’ll try to do that in a follow-up PR if I’ve got time. Filed rdar://102853071 for now.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Pull request merged**